### ISSUE TITLE
Shape inference on the Torchvision’s Mask R-CNN causes a segmentation fault

### ISSUE THREAD

**Issue by przybyszewskiw** [onnx/onnx#2965]
# Bug Report

### Is the issue related to model conversion?
Probably not.

### Describe the bug
When I try to run shape inference on the Torchvision's Mask R-CNN it causes a segfault. However, `check_model` doesn't return any warning/exception.

### System information
- OS Platform and Distribution (*e.g. Linux Ubuntu 16.04*): Linux Ubuntu 18.04.4
- ONNX version (*e.g. 1.7*):  1.6.0
- Python version: 3.6.10
- PyTorch version: 1.6.0a0+9907a3e
- Torchvision version: 0.7.0a0
- GCC/Compiler version (if compiling from source): N/A
- CMake version: 3.14.0
- Protobuf version: 3.12.2
- Visual Studio version (if applicable): N/A

### Reproduction instructions
```
import torch
import torchvision
import onnx

model = torchvision.models.detection.maskrcnn_resnet50_fpn(pretrained=True)
model.eval()
x = [torch.rand(3, 800, 800)]
torch.onnx.export(model, x, "mask_rcnn.onnx", opset_version = 11)
onnx_model = onnx.load("mask_rcnn.onnx")
onnx.checker.check_model(onnx_model)
onnx.shape_inference.infer_shapes(onnx_model)
```

### Expected behavior
Shape inference doesn't cause any errors.

**Comment by jcwchen** [comment]
Hi @przybyszewskiw,
Sorry that current shape inference does not catch the exception properly. 
This segfault bug will be fixed in this PR: https://github.com/onnx/onnx/pull/2783.

After applying this PR, I tried to check this model with `onnx.checker.check_model('mask_rcnn.onnx', full_check=True)` and got the error as follows:
```
Invalid tensor data type.
```

For detailed error messages, I "shape_inference" the model and got the error as follows:
```
Shape inference error(s): (op_type:Gather, name:Gather_2359): [TypeInferenceError] Input 0 expected to have tensor type
(op_type:Gather, name:Gather_2360): [TypeInferenceError] Input 0 expected to have tensor type
(op_type:Slice, name:Slice_2389): [TypeInferenceError] Input 0 expected to have tensor type
(op_type:Slice, name:Slice_2398): [TypeInferenceError] Input 0 expected to have tensor type

```

**Comment by przybyszewskiw** [comment]
Hi @jcwchen,
Thank you for your reply. I extended my script, so now it performs an inference in the `onnxruntime` as well:
```
import torch
import torchvision
import onnxruntime
import numpy as np
import requests
from io import BytesIO
from PIL import Image

def to_numpy(tensor):
    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

def get_image(url):
    response = requests.get(url)
    pil_image = Image.open(BytesIO(response.content)).convert("RGB")
    return torch.tensor(np.array(pil_image)).permute(2, 1, 0).to(dtype=torch.float)/255

model = torchvision.models.detection.maskrcnn_resnet50_fpn(pretrained=True)
model.eval()

url = "http://farm3.staticflickr.com/2469/3915380994_2e611b1779_z.jpg"
x = get_image(url)

torch_out = model([x])
torch.onnx.export(model, [x], "mask_rcnn.onnx", opset_version=11, input_names = ['x'])

ort_session = onnxruntime.InferenceSession("mask_rcnn.onnx")
ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
ort_outs = ort_session.run(None, ort_inputs)

for i, key in enumerate(torch_out[0].keys()):
    np.testing.assert_allclose(to_numpy(torch_out[0][key]), ort_outs[i], rtol=1e-03, atol=2e-05)

print("Exported model passed the test!")
```
It seems like the exported model does the inference correctly, even though the type inference fails. Could you explain this inconsistency?

**Comment by jcwchen** [comment]
Can you specify what the type inference failure is? As far I know, type inference of onnxruntime must finish successfully so it should print error messages if any.

**Comment by przybyszewskiw** [comment]
There is no error when I run the script above. I mentioned the type inference failure you wrote above. How is it that I can run inference on my model without any problem, but when I run shape_inference it segfaults because of TypeInferenceErrors?

**Comment by jcwchen** [comment]
I think the reason is onnxruntime has its own shape_inference and it only uses part of onnx.shape_inference (node-level). If you use shape inference from onnxruntime successfully, the type inference of your model should be done.

**Comment by gramalingam** [comment]
This is a bit strange. As far as I know, the inference used by onnxruntime is the same regardless of whether you use the python or C/C++ API. There is a slightly different implementation of the inference in onnxruntime and ONNX. But even here, if the behavior is different, then one of them needs to be fixed. This seems worth investigating.

**Comment by jcwchen** [comment]
Thanks @gramalingam for providing a PR to fix the type inference! This issue should be resolved by that PR. @przybyszewskiw if you still encounter the same error, please reopen it. Thank you.

**Comment by jcwchen** [comment]
(This might be an onnxruntime issue)
@przybyszewskiw One more thing: the gap between Python API and C API for onnxruntime. This is not supposed to happen.
 
1. Could you try the following code on your end? Let's see whether it will encounter the same error as C API.
```
import onnxruntime as ort
so = onnxrt.SessionOptions()
so.graph_optimization_level = onnxrt.GraphOptimizationLevel.ORT_ENABLE_BASIC
onnxrt.InferenceSession("mask_rcnn.onnx") 
```
2. Did you build C API and Python API with the same source code? What is your onnxruntime and onnx version?

Thank you for the help.

**Comment by przybyszewskiw** [comment]
@jcwchen after the PR I don't have any problems with the `infere_shapes` function in Python API. Thanks for that!

When it comes to this gap between APIs:
1. I don't encounter any bug while running your code - modifying `SessionOptions` doesn't change anything;
2. Yes, my C and Python APIs are built with the same source code. My onnx version is 1.6.0 and onnxruntime version is 1.3.0.

Can you reproduce that bug? Should I start a new issue for it?

**Comment by jcwchen** [comment]
Thanks for the answer. My previous testing versions of onnxruntime and onnx are different from yours and I will try to reproduce it. Could you try to test the same model with  newer onnxruntime 1.4.0 and onnx 1.7.0 if possible?
If the Python API and C API still behave differently, feel free to open an issue under onnxruntime. Let's track this problem there. Thanks!

**Comment by przybyszewskiw** [comment]
It looks like when I use onnx 1.7.0 and onnxruntime 1.4.0, model loads correctly in both APIs. Thanks for your help!
